// @translate-tools/core/translators/FakeTranslator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FakeTranslator = exports.ErrorFakeTranslator = void 0;

var _BaseTranslator = require("../util/BaseTranslator");

var __extends = void 0 && (void 0).__extends || function () {
  var extendStatics = function (d, b) {
    extendStatics = Object.setPrototypeOf || {
      __proto__: []
    } instanceof Array && function (d, b) {
      d.__proto__ = b;
    } || function (d, b) {
      for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p];
    };

    return extendStatics(d, b);
  };

  return function (d, b) {
    if (typeof b !== "function" && b !== null) throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
    extendStatics(d, b);

    function __() {
      this.constructor = d;
    }

    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
  };
}();

var __awaiter = void 0 && (void 0).__awaiter || function (thisArg, _arguments, P, generator) {
  function adopt(value) {
    return value instanceof P ? value : new P(function (resolve) {
      resolve(value);
    });
  }

  return new (P || (P = Promise))(function (resolve, reject) {
    function fulfilled(value) {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    }

    function rejected(value) {
      try {
        step(generator["throw"](value));
      } catch (e) {
        reject(e);
      }
    }

    function step(result) {
      result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
    }

    step((generator = generator.apply(thisArg, _arguments || [])).next());
  });
};

var __generator = void 0 && (void 0).__generator || function (thisArg, body) {
  var _ = {
    label: 0,
    sent: function () {
      if (t[0] & 1) throw t[1];
      return t[1];
    },
    trys: [],
    ops: []
  },
      f,
      y,
      t,
      g;
  return g = {
    next: verb(0),
    "throw": verb(1),
    "return": verb(2)
  }, typeof Symbol === "function" && (g[Symbol.iterator] = function () {
    return this;
  }), g;

  function verb(n) {
    return function (v) {
      return step([n, v]);
    };
  }

  function step(op) {
    if (f) throw new TypeError("Generator is already executing.");

    while (_) try {
      if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
      if (y = 0, t) op = [op[0] & 2, t.value];

      switch (op[0]) {
        case 0:
        case 1:
          t = op;
          break;

        case 4:
          _.label++;
          return {
            value: op[1],
            done: false
          };

        case 5:
          _.label++;
          y = op[1];
          op = [0];
          continue;

        case 7:
          op = _.ops.pop();

          _.trys.pop();

          continue;

        default:
          if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
            _ = 0;
            continue;
          }

          if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
            _.label = op[1];
            break;
          }

          if (op[0] === 6 && _.label < t[1]) {
            _.label = t[1];
            t = op;
            break;
          }

          if (t && _.label < t[2]) {
            _.label = t[2];

            _.ops.push(op);

            break;
          }

          if (t[2]) _.ops.pop();

          _.trys.pop();

          continue;
      }

      op = body.call(thisArg, _);
    } catch (e) {
      op = [6, e];
      y = 0;
    } finally {
      f = t = 0;
    }

    if (op[0] & 5) throw op[1];
    return {
      value: op[0] ? op[1] : void 0,
      done: true
    };
  }
};

/**
 * Fake translator for use in tests and debug
 */
var FakeTranslator =
/** @class */
function (_super) {
  __extends(FakeTranslator, _super);

  function FakeTranslator() {
    return _super !== null && _super.apply(this, arguments) || this;
  }

  FakeTranslator.isSupportedAutoFrom = function () {
    return false;
  };

  FakeTranslator.getSupportedLanguages = function () {
    return ['ru', 'en', 'de', 'ja'];
  };

  FakeTranslator.prototype.getLengthLimit = function () {
    return 3000;
  };

  FakeTranslator.prototype.getRequestsTimeout = function () {
    return 10;
  };

  FakeTranslator.prototype.checkDirection = function (from, to) {
    return from == 'ru' && to == 'ja' ? false : true;
  };

  FakeTranslator.prototype.translate = function (text, from, to) {
    var delay = this.options.delay === undefined ? 0 : this.options.delay === 'random' ? Math.floor(Math.random() * 1000) : this.options.delay;
    return new Promise(function (resolve) {
      setTimeout(function () {
        return resolve("*[".concat(from, "-").concat(to, "]") + text);
      }, delay);
    });
  };

  FakeTranslator.prototype.translateBatch = function (text, from, to) {
    var _this = this;

    return Promise.all(text.map(function (i) {
      return _this.translate(i, from, to).catch(function () {
        return null;
      });
    }));
  };

  FakeTranslator.translatorName = 'FakeTranslator';
  return FakeTranslator;
}(_BaseTranslator.BaseTranslator);

exports.FakeTranslator = FakeTranslator;

/**
 * Fake translator which always throw error for use in tests and debug
 */
var ErrorFakeTranslator =
/** @class */
function (_super) {
  __extends(ErrorFakeTranslator, _super);

  function ErrorFakeTranslator() {
    return _super !== null && _super.apply(this, arguments) || this;
  }

  ErrorFakeTranslator.prototype.translate = function (_text, _from, _to) {
    return __awaiter(this, void 0, void 0, function () {
      return __generator(this, function (_a) {
        throw new Error('Fake error for translate method');
      });
    });
  };

  ErrorFakeTranslator.prototype.translateBatch = function (_text, _from, _to) {
    return __awaiter(this, void 0, void 0, function () {
      return __generator(this, function (_a) {
        throw new Error('Fake error for translateBatch method');
      });
    });
  };

  ErrorFakeTranslator.translatorName = 'FakeTranslator';
  return ErrorFakeTranslator;
}(FakeTranslator);

exports.ErrorFakeTranslator = ErrorFakeTranslator;
