// @translate-tools/core/tts/GoogleTTS/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GoogleTTS = void 0;

var _axios = _interopRequireDefault(require("axios"));

var _buffers = require("../../util/buffers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var GoogleTTS =
/** @class */
function () {
  function GoogleTTS() {}

  GoogleTTS.prototype.getAudioBuffer = function (text, language) {
    var url = "https://translate.google.com/translate_tts?ie=UTF-8&tl=".concat(language, "&client=dict-chrome-ex&ttsspeed=0.5&q=") + encodeURIComponent(text);
    return (0, _axios.default)({
      url: url,
      method: 'GET',
      responseType: 'arraybuffer'
    }).then(function (_a) {
      var data = _a.data;
      var buffer;

      if (typeof Buffer !== 'undefined' && data instanceof Buffer) {
        buffer = (0, _buffers.bufferToArrayBuffer)(data);
      } else if (data instanceof ArrayBuffer) {
        buffer = data;
      } else {
        throw new TypeError('Unexpected response');
      }

      return {
        type: 'audio/mpeg',
        buffer: buffer
      };
    });
  };

  GoogleTTS.getSupportedLanguages = function () {
    // prettier-ignore
    return ["af", "sq", "am", "ar", "hy", "as", "ay", "az", "bm", "eu", "be", "bn", "bho", "bs", "bg", "ca", "ceb", "ny", "zh", "zh_HANT", "co", "hr", "cs", "da", "dv", "doi", "nl", "en", "eo", "et", "ee", "tl", "fi", "fr", "fy", "gl", "ka", "de", "el", "gn", "gu", "ht", "ha", "haw", "iw", "hi", "hmn", "hu", "is", "ig", "ilo", "id", "ga", "it", "ja", "jw", "kn", "kk", "km", "rw", "gom", "ko", "kri", "ku", "ckb", "ky", "lo", "la", "lv", "ln", "lt", "lg", "lb", "mk", "mai", "mg", "ms", "ml", "mt", "mi", "mr", "mni-Mtei", "lus", "mn", "my", "ne", "no", "or", "om", "ps", "fa", "pl", "pt", "pa", "qu", "ro", "ru", "sm", "sa", "gd", "nso", "sr", "st", "sn", "sd", "si", "sk", "sl", "so", "es", "su", "sw", "sv", "tg", "ta", "tt", "te", "th", "ti", "ts", "tr", "tk", "ak", "uk", "ur", "ug", "uz", "vi", "cy", "xh", "yi", "yo", "zu"];
  };

  return GoogleTTS;
}();

exports.GoogleTTS = GoogleTTS;
